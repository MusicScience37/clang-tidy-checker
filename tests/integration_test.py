"""Integration tests."""

import pathlib
import subprocess
import typing

import approvaltests
import pytest

from .path_scrubber import PATH_SCRUBBER
from .warning_count_scrubber import WARNING_COUNT_SCRUBBER

THIS_DIR = pathlib.Path(__file__).absolute().parent


async def execute(command: typing.List[str], *, cwd: str, repeat: bool = False) -> None:
    """Execute a command."""

    repetitions = 1
    if repeat:
        repetitions = 3

    for _ in range(repetitions):
        result = subprocess.run(
            command,
            cwd=cwd,
            check=False,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
        )

    approvaltests.approvals.verify(
        f"""exit_code: {result.returncode}
stdout:
{result.stdout.decode("utf8")}
stderr:
{result.stderr.decode("utf8")}
""",
        options=approvaltests.Options().with_scrubber(
            approvaltests.scrubbers.combine_scrubbers(
                PATH_SCRUBBER, WARNING_COUNT_SCRUBBER
            )
        ),
    )


@pytest.mark.asyncio
async def test_help(sample_proj_no_error: pathlib.Path):
    """Test of help message."""

    await execute(
        ["clang-tidy-checker", "--help"],
        cwd=str(sample_proj_no_error),
    )


@pytest.mark.asyncio
async def test_config(sample_proj_no_error: pathlib.Path):
    """Test of config option."""

    await execute(
        [
            "clang-tidy-checker",
            "--no-ascii",
            "--config",
            str(sample_proj_no_error.parent / "configs" / "main_only.yaml"),
        ],
        cwd=str(sample_proj_no_error),
    )


@pytest.mark.asyncio
async def test_build_dir(sample_proj_no_error: pathlib.Path):
    """Test of build_dir option."""

    await execute(
        ["clang-tidy-checker", "--no-ascii", "--build_dir", "."],
        cwd=str(sample_proj_no_error),
    )


@pytest.mark.asyncio
async def test_check_proj_no_error(sample_proj_no_error: pathlib.Path):
    """Test of checking project without errors."""

    await execute(
        ["clang-tidy-checker", "--no-ascii"],
        cwd=str(sample_proj_no_error),
    )


@pytest.mark.asyncio
async def test_check_proj_warning(sample_proj_warning: pathlib.Path):
    """Test of checking project with a warnings."""

    await execute(
        ["clang-tidy-checker", "--no-ascii"],
        cwd=str(sample_proj_warning),
    )


@pytest.mark.asyncio
async def test_check_proj_error(sample_proj_error: pathlib.Path):
    """Test of checking project with an error."""

    await execute(
        ["clang-tidy-checker", "--no-ascii"],
        cwd=str(sample_proj_error),
    )


@pytest.mark.asyncio
async def test_extra_args(sample_proj_warning: pathlib.Path):
    """Test of extra arguments to clang-tidy."""

    await execute(
        [
            "clang-tidy-checker",
            "--no-ascii",
            "--extra_arg",
            "--extra-arg=-any",
        ],
        cwd=str(sample_proj_warning),
    )


@pytest.mark.asyncio
async def test_cache_dir(sample_proj_warning: pathlib.Path):
    """Test of cache directory."""

    await execute(
        [
            "clang-tidy-checker",
            "--no-ascii",
            "--cache_dir",
            str(THIS_DIR.parent / ".clang-tidy-cache"),
        ],
        cwd=str(sample_proj_warning),
        repeat=True,
    )
