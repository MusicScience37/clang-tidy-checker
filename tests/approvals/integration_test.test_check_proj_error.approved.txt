exit_code: 1
stdout:

stderr:
Check of <root-dir>/sample_inputs/proj_error/src/main.cpp finished with exit code 0.
Check of <root-dir>/sample_inputs/proj_error/src/sample_function.cpp finished with exit code 1.
<root-dir>/sample_inputs/proj_error/src/sample_function.cpp:3:26: error: use of undeclared identifier 'std' [clang-diagnostic-error]
    3 | void sample_function() { std::cout << "Test" << std::endl; }
      |                          ^
<root-dir>/sample_inputs/proj_error/src/sample_function.cpp:3:49: error: use of undeclared identifier 'std' [clang-diagnostic-error]
    3 | void sample_function() { std::cout << "Test" << std::endl; }
      |                                                 ^

2 errors generated.
Error while processing <root-dir>/sample_inputs/proj_error/src/sample_function.cpp.

Some errors detected.

